### PR TITLE
Fix Resume Course 404 issue

### PR DESCRIPTION
**Pull request by shadinaif**
## Change description

Fix Resume Course 404 issue. Simulate https://github.com/openedx/edx-platform/pull/27046 without upgrading completion-package

## Type of change
- [x] Bug fix (fixes an issue)
- [ ] New feature (adds functionality)

## Related issues

Related to: https://appsembler.atlassian.net/browse/RED-3276

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

**Review by OmarIthawi — CHANGES_REQUESTED** [review]
Thanks @shadinaif. Please change to the following:

 - Convert into a dedicated helper so the code isn't stray and we can tell it's our code: `appsembler.future_releases_hacks.helpers` module.
 - Import the helper in the `student/helpers.py` module
 - Add tests to the helper.
 - Add this PR to the list of tech-debt issues we have until we upgrade.

**Comment by shadinaif** [comment]
Thank you @OmarIthawi , please take a look now

File: openedx/core/djangoapps/appsembler/future_releases_hacks/helpers.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    """
+    Helper to simulate completion.utilities.get_key_to_last_completed_block version 4. We need this version to fix
+    a bug in dashboard Resume Course button, but we cannot upgrade the package within Juniper. RED-3276
+    """
+    last_completed_block = BlockCompletion.get_latest_block_completed(user, context_key)
+
+    if last_completed_block is not None:
+        return last_completed_block.full_block_key
+
+    raise UnavailableCompletionData(context_key)
```
Comment (OmarIthawi) [inline openedx/core/djangoapps/appsembler/future_releases_hacks/helpers.py:19]:
```py
def url_to_block():
  try:
      block_data = get_course_blocks(user, block_key)
  except UsageKeyNotInBlockStructure:
      url_to_block = ''
  else:
      if block_key in block_data:
          url_to_block = reverse(
              'jump_to',
              kwargs={'course_id': enrollment.course_id, 'location': block_key}
          )
      else:
          url_to_block = ''
```

one more method please. This is going to help a lot during a release merge. 

it's very desirable to have as little change to upstream code as possible.

File: common/djangoapps/student/helpers.py [hunk 2]
```diff
@@ -11,13 +11,13 @@
...
 
 from completion.exceptions import UnavailableCompletionData
-from completion.utilities import get_key_to_last_completed_block
 from django.conf import settings
 from django.contrib.auth import load_backend
 from django.contrib.auth.models import User
 from django.core.exceptions import PermissionDenied
 from django.core.validators import ValidationError
 from django.db import IntegrityError, transaction
+
```
Comment (OmarIthawi) [inline common/djangoapps/student/helpers.py:20]: git noise

```suggestion
```

File: common/djangoapps/student/helpers.py [hunk 3]
```diff
@@ -27,12 +27,15 @@
...
 from lms.djangoapps.certificates.models import CertificateStatuses, certificate_status_for_student
+from lms.djangoapps.course_blocks.api import get_course_blocks
 from lms.djangoapps.grades.api import CourseGradeFactory
 from lms.djangoapps.verify_student.models import VerificationDeadline
 from lms.djangoapps.verify_student.services import IDVerificationService
 from lms.djangoapps.verify_student.utils import is_verification_expiring_soon, verification_for_datetime
+from openedx.core.djangoapps.appsembler.future_releases_hacks.helpers import get_key_to_last_completed_block
 from openedx.core.djangoapps.ace_common.template_context import get_base_template_context
 from openedx.core.djangoapps.certificates.api import certificates_viewable_for_course
+from openedx.core.djangoapps.content.block_structure.exceptions import UsageKeyNotInBlockStructure
```
Comment (OmarIthawi) [inline common/djangoapps/student/helpers.py:38]: Keep our imports at the bottom of upstream imports. This reduces conflicts by a ton.

**Review by OmarIthawi — COMMENTED** [review]
Thanks @shadinaif, one more clean up please and it should be okay.

File: common/djangoapps/student/helpers.py (hunk 2 above)
Comment (shadinaif) [reply common/djangoapps/student/helpers.py]: agree

**Comment by shadinaif** [comment]
Thank you @OmarIthawi , and sorry that I didn't mention that after the first review: I removed my code and cherry-picked from upstream https://github.com/openedx/edx-platform/commit/f4974aebd52c08167f74f12769ef8f18e289627b . So, the new change request will create future conflicts, unlike the thing you're aiming for :)

**Review by OmarIthawi — APPROVED** [review]
> Thank you @OmarIthawi , and sorry that I didn't mention that after the first review: I removed my code and cherry-picked from upstream [openedx@f4974ae](https://github.com/openedx/edx-platform/commit/f4974aebd52c08167f74f12769ef8f18e289627b) . So, the new change request will create future conflicts, unlike the thing you're aiming for :)

Shocking 😆

**Pull request merged**